pkg/odo/cli/init/init.go

```diff
@@ -0,0 +1,306 @@
+		}
+	}()
+
+	destDevfile := filepath.Join(o.destDir, "devfile.yaml")
+	if o.InitParams.DevfilePath != "" {
+		err = o.downloadDirect(o.InitParams.DevfilePath, destDevfile)
+	} else {
+		err = o.downloadRegistry(o.InitParams.DevfileRegistry, o.InitParams.Devfile, o.destDir)
+	}
+	if err != nil {
+		return fmt.Errorf("Unable to download devfile: %w", err)
+	}
+
+	devfileObj, err := devfile.ParseAndValidateFromFile(destDevfile)
+	if err != nil {
+		return err
+	}
+
+	if o.InitParams.Starter != "" {
+		// WARNING: this will remove all the content of the destination directory, ie the devfile.yaml file
+		err = o.downloadStarterProject(devfileObj, o.InitParams.Starter, o.destDir)
+		if err != nil {
+			return fmt.Errorf("unable to download starter project %q: %w", o.InitParams.Starter, err)
+		}
+		starterDownloaded = true
+
+		// in case the starter project contains a devfile, read it again
+		if _, err = o.fsys.Stat(destDevfile); err == nil {
+			devfileObj, err = devfile.ParseAndValidateFromFile(destDevfile)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Set the name in the devfile *AND* writes the devfile back to the disk in case
+	// it has been removed and not replaced by the starter project
+	err = devfileObj.SetMetadataName(o.InitParams.Name)
+	if err != nil {
+		return fmt.Errorf("Failed to update the devfile's name: %w", err)
+	}
+
+	log.Italicf(`Your new component %q is ready in the current directory.
+To start editing your component, use "odo dev" and open this folder in your favorite IDE.
+Changes will be directly reflected on the cluster.
+To deploy your component to a cluster use “odo deploy”.`, o.InitParams.Name)
+
+	return nil
+}
+
+// downloadDirect downloads a devfile at the provided URL and saves it in dest
+func (o *InitOptions) downloadDirect(URL string, dest string) error {
+	parsedURL, err := url.Parse(URL)
+	if err != nil {
+		return err
+	}
+	if strings.HasPrefix(parsedURL.Scheme, "http") {
+		downloadSpinner := log.Spinnerf("Downloading devfile from %q", URL)
+		defer downloadSpinner.End(false)
+		params := util.HTTPRequestParams{
+			URL: URL,
+		}
+		devfileData, err := util.DownloadFileInMemory(params)
+		if err != nil {
+			return err
+		}
+		err = o.fsys.WriteFile(dest, devfileData, 0644)
+		if err != nil {
+			return err
+		}
+		downloadSpinner.End(true)
+	} else {
+		downloadSpinner := log.Spinnerf("Copying devfile from %q", URL)
+		defer downloadSpinner.End(false)
+		content, err := o.fsys.ReadFile(URL)
+		if err != nil {
+			return err
+		}
+		info, err := o.fsys.Stat(URL)
+		if err != nil {
+			return err
+		}
+		err = o.fsys.WriteFile(dest, content, info.Mode().Perm())
+		if err != nil {
+			return err
+		}
+		downloadSpinner.End(true)
+	}
+
+	return nil
+}
+
+// downloadRegistry downloads a devfile from the provided registry and saves it in dest
+func (o *InitOptions) downloadRegistry(registryName string, devfile string, dest string) error {
```

Comment:
IMO, `downloadFromRegistry` is more appropriate
```suggestion
// downloadFromRegistry downloads a devfile from the provided registry and saves it in dest
func (o *InitOptions) downloadFromRegistry(registryName string, devfile string, dest string) error {
```